proposals/6/README.md

```diff
@@ -0,0 +1,123 @@
+but simply query the Kubernetes API and, depending on its state, create/update other Kubernetes API objects.
+For instance long running logic can be moved into a Pod that is created by a controller.
+It should also not actively wait for the created/updated API objects to reach a certain state. Instead it should be ensured that these objects are watched
+so that a new reconcile request is created whenever such an object changes.  
+
+This architecture allows to hide complex imperatively coded processes behind a declarative interface that is technology-agnostic and easier to understand for humans.  
+
+### 2.2 The Operator Pattern
+
+The Operator Pattern components that encorporate expert knowledge for a particular application in order to automate its operation.
+Therefore these components are called operators.
+An operator basically consists of one or many CRDs and a controller for each.
+(Following the K8s conventions each git repository that specifies K8s APIs should maintain its CRDs at `deploy/crds` and the corresponding go types at `pkg/apis`.)
+
+## 3. Principles
+
+This section proposes some principles we should embrace. Most of them are valid beyond Kubernetes.
+
+* Inversion of Control: Build passive components whose processes are triggered by events or changes in watched API resources.
+  * Build components in a way so that they can be developed and tested ideally isolated from the rest
+  * Support running as much as possible in minikube to lower the contribution and experimentation barrier.
+* Modularity:
+  * Design CRDs modularly so that they can be reused/combined with other components.
+* Resilience: Favour small pieces of business logic that can be run in a controller's reconcile loop over long-running blocking code with active retry implementations.
+* Performance/Concurrency: Don't run long blocking code within a controller's reconcile loop (otherwise reduces concurrency).
+* Performance: Favour batched API changes (kubectl apply -f -) that don't take time for resources that haven't changed over sequential long running pipelines that do a lot even if it is not necessary (applies to GitOps in particular).
+* Simplicity/Usability: Favour declarative k8s-native interfaces over imperative code.
+* Usability/Understanding/Performance: Design denormalized CRDs (corresponds to document DB principles): When a controller creates other resources for a given resource it should continuously watch and write back their state to the owning resource so that users have all relevant status information at hand by querying the owning resource only.
+  * Correspondingly avoid joining API resources - at least when listing them.
+* Security: Don't mix security concerns. Think of how different security-related responsibilities can be separated before implementing new features. Model CRDs so that K8s' RBAC can be leveraged.
+
+## 4. External interface(s) & user load
```

Comment:
Deane recommends to call them "public" interface and encourage users to use these only instead of relying on potentially more unstable internal k8s APIs.